presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java

```diff
@@ -177,13 +193,13 @@ public boolean isFinished()
     public boolean needsInput()
     {
         updateBlockedIfNecessary();
-        return state == State.RUNNING && blocked == NOT_BLOCKED;
+        return state == State.RUNNING && blocked == NOT_BLOCKED && statisticAggregation.needsInput();
     }
 
     private void updateBlockedIfNecessary()
     {
         if (blocked != NOT_BLOCKED && blocked.isDone()) {
-            blocked = NOT_BLOCKED;
+            blocked = statisticAggregation.isBlocked();
```

Comment:
Yeah, you are right.

I should've spent more time analyzing the interactions. Trying to solve the race with calling `isBlocked` + `allAsList`. It seems to be that now it should be fine. But i would ask you to verify it carefully as well. I could've made a mistake again.

Reply:
`allAsList` has its own implications that we need to be aware of -- every time it's called, it adds a listener to the underlying futures, leaking memory.
In `Driver`, we address this with being less "functional", caching the future (`com.facebook.presto.operator.Driver#driverBlockedFuture`). If you wish, I can think about this more.

Reply:
I'm well aware of that issue. Previously we had problems because we were creating a new `allAsList` every time the `isBlocked` method was called. I'm creating an `allAsList` future on `addPage` or `finish`, when the state is actually changing.